<a href="https://colab.research.google.com/github/AnnaShtyn/python_for_hw_tasks/blob/main/Anna_Shtyn_HW_15_2_%D0%9F%D0%B5%D1%80%D0%B5%D0%B2%D1%96%D1%80%D0%BA%D0%B0_%D1%81%D1%82%D0%B0%D1%82%D0%B8c%D1%82%D0%B8%D1%87%D0%BD%D0%B8%D1%85_%D0%B3%D1%96%D0%BF%D0%BE%D1%82%D0%B5%D0%B7_Z_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [27]:
import pandas as pd
import numpy as np
from scipy import stats

data = pd.read_csv('drive/MyDrive/data.csv')

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [28]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. (7 балів)
Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [29]:
# 1.
sample_mean = np.mean(data)
population_mean = 20
population_std = np.std(data, ddof=1)
sample_size = data.size
alpha = 0.05

print(f'x¯={sample_mean}\nμ0={population_mean}\nσ={population_std}\nn={sample_size}\nα={alpha}')

x¯=19.37755319670441
μ0=20
σ=6.25356123240434
n=100
α=0.05


2.`Ho: μ = 20`

`Ha: μ > 20`

Правостороній тест

In [32]:
# 3.
SE = population_std / np.sqrt(sample_size)

print(f'SE = {SE}')

SE = 0.625356123240434


In [33]:
# 4.
z_stat = (sample_mean - population_mean) / SE
print(f'z_stat = {z_stat}')

z_stat = -0.995347738933505


In [35]:
# 5.
z_crit = stats.norm.ppf(1-alpha)
print(f'z_crit = {z_crit}')

z_crit = 1.6448536269514722


In [36]:
# 6.
p_value = 1 - stats.norm.cdf(z_stat)
print(f'p_value = {p_value}')

p_value = 0.8402164165464012


7.
Отже, порівнявши обраховану  z-статистику з критичним z-значенням (|-0.995| < 1.645 = |z_stat| < z_crit) бачимо, що z_stat менша. Це означає, що в нас немає підстав для відхилення нульової гіпотези (немає підстав, щоб прийняти гіпотезу На).
  
  P-value також значно перевищує рівень значущості (0,84 > 0,05).

**Отже, ми ніяк не можемо відхилити Но гіпотезу, оскільки в нас недостатньо доказів в правдивості гіпотези На.**


**Завдання 2.** (3 бали)
Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

In [49]:
from statsmodels.stats import weightstats as stests
z_stat_sm, p_value_sm = stests.ztest(data,
                         value = population_mean,
                         alternative = 'larger')
print(f'Z-stat={z_stat_sm}\nP-value={p_value_sm}')

# В нас недостатньо підстав щоб прийняти гіпотезу На, тому підтвердити зростання
#  сер.чеку ми не можемо.
# P-value = 0.84, тобто ймовірність отримати середній чек в 19,38 дол.,припускаючи,
#  що середній чек становить 20 дол (гіпотеза Но) становить аж 84%. Таким чином,
#  вагомих доказів в підтримку гіпотези про зростання сер.чеку (На) в нас немає.

z_test1 = [z_stat, p_value]
z_test2 = [z_stat_sm, p_value_sm]
print(f"\n{'👍Показники збігаються за 2ома розвязками' if np.allclose(z_test1, z_test2) == True else 'Уппс'}")

Z-stat=-0.9953477389335053
P-value=0.8402164165464012

👍Показники збігаються за 2ома розвязками
